### PR TITLE
173 Cage Rework

### LINKED ISSUE #835: everyone in radius blinks instantly after 173 cage opens
#### Description of issue
title

#### Difference between expected and actual behavior
everyone should not blink instantly

#### Steps to reproduce
```
Jasper Coldsmith attempts to open SCP-173 Cage.
Jasper Coldsmith opens SCP-173 Cage!
Checkers Shafer blinks.
Jasper Coldsmith blinks.
Chad Blackburn blinks.
Unknown (as Zachary Hooker) blinks.
SCP-173 snaps Chad Blackburn's neck!
```

### PR DESCRIPTION
**Pull request by TheDarkElites**
## About the Pull Request
173 has new cage mechanics! Wooh!
173 now has to damage its cage five times before escaping.
It can only damage its cage when not being watched.
The cage now needs to be repaired after being broken by 173.

## Why It's Good For The Game
Usually, when something that can only move when not being watched is in a cage the best way to contain it is to look at it, not move the cage around a bunch.

Also fixes #835

## Changelog

:cl:
add: Changed 173 cage mechanics
imageadd: New cage images so 173 can actually fit.
/:cl:

### CODE REVIEW

**Comment by TheDarkElites** [comment]
Put this up early cause I had to get updated dev-branch so sshhhh don't look at anything

File: code/modules/SCP/SCPs/SCP-173.dm [hunk 1]
```diff
@@ -91,7 +91,19 @@ GLOBAL_LIST_EMPTY(scp173s)
...
+			if(snap_cooldown > world.time)
+				var/obj/structure/scp173_cage/cage = A
+				cage.attack_hand(src)
+				snap_cooldown = world.time + snap_cooldown_time
+			else
+				to_chat(src, "<span class='warning'>You can't attack yet.</span>")
+		else if(IsBeingWatched() && istype(A, /obj/structure/scp173_cage) && loc == "SCP-173 Containment Cage")
+			to_chat(src, "<span class='warning'>You can't attack the cage while they're looking!</span>")
+		else
+			return
```
Comment (MACIEKBAKI) [inline code/modules/SCP/SCPs/SCP-173.dm]:
```suggestion
	if(incapacitated()) // We can't do anything while caged
		if(istype(A, /obj/structure/scp173_cage) && loc == A)
			var/obj/structure/scp173_cage/cage = A
			if(!IsBeingWatched()) //Can only attack the cage when not being watched and inside the cage
				if(snap_cooldown > world.time)
					cage.attack_hand(src)
					snap_cooldown = world.time + snap_cooldown_time
				else
					to_chat(src, SPAN_WARNING("You can't attack yet."))
					return
			else
				to_chat(src, SPAN_WARNING("You can't attack the cage while they're looking!"))
				return
		else
			return
	if(IsBeingWatched())// We can't do anything while being watched
```

File: code/modules/SCP/SCPs/SCP-173.dm [hunk 2]
```diff
@@ -343,14 +355,23 @@ GLOBAL_LIST_EMPTY(scp173s)
...
-// the cage
+//
+// 173 Cage
+//
+
 /obj/structure/scp173_cage
 	icon = 'icons/SCP/cage.dmi'
-	icon_state = "1"
-	name = "Empty SCP-173 Cage"
+	icon_state = "2"
```
Comment (MACIEKBAKI) [inline code/modules/SCP/SCPs/SCP-173.dm]: what are these icon state names, who named them, jeez (not issue with pr, issue with whoever named the sprites)

File: code/modules/SCP/SCPs/SCP-173.dm [hunk 3]
```diff
@@ -359,30 +380,38 @@ GLOBAL_LIST_EMPTY(scp173s)
...
-		visible_message(SPAN_DANGER("[H] opens \the [src]!"))
+	if(istype(A, /mob/living/carbon/human))
+		var/mob/living/carbon/human/H = A
+		visible_message(SPAN_WARNING("[H] attempts to open \the [src]."))
+		if(do_after(H, 5 SECONDS, src))
+			visible_message(SPAN_DANGER("[H] opens \the [src]!"))
+			ReleaseContents()
+	else if(istype(A, /mob/living/scp_173))
+		var/mob/living/scp_173/S = A
 		ReleaseContents()
```
Comment (MACIEKBAKI) [inline code/modules/SCP/SCPs/SCP-173.dm]: hahahaha 173s can release other 173s 
tho why is `var/mob/living/scp_173/S = A` needed?
also I'd add a check to make sure the 173 that is releasing is *not* being watched / not caged itself

**Comment by TheDarkElites** [comment]
I would just like to once again point to my PR and say this isn't done and WIP. I had to put it up to marge dev branch with mine, so the first commit will look weird.

File: code/modules/SCP/SCPs/SCP-173.dm (hunk 3 above)
Comment (TheDarkElites) [reply code/modules/SCP/SCPs/SCP-173.dm]: that isn't done its just for testing purposes, currently working on it

**Comment by MACIEKBAKI** [comment]
((( i did not notice the fact its a draft )))

**Comment by TheDarkElites** [comment]
> ((( i did not notice the fact its a draft )))

:)

File: code/modules/SCP/SCPs/SCP-173.dm [hunk 4]
```diff
@@ -91,7 +96,20 @@ GLOBAL_LIST_EMPTY(scp173s)
...
 
 /mob/living/scp_173/UnarmedAttack(atom/A)
-	if(IsBeingWatched() || incapacitated()) // We can't do anything while being watched
+	if(is_caged) // We can't do anything while caged
+		if(!IsBeingWatched() && istype(A, /obj/structure/scp173_cage)) //Can only attack the cage when not being watched and inside the cage
+			if(snap_cooldown <= world.time)
+				var/obj/structure/scp173_cage/cage = A
+				cage.attack_hand(src)
+				snap_cooldown = world.time + snap_cooldown_time
+				to_chat(src, "<span class='warning'>You damage the cage.</span>")
```
Comment (aidan-bowman) [inline code/modules/SCP/SCPs/SCP-173.dm]: Use SPAN_WARNING and the other defines

**Comment by TheDarkElites** [comment]
@cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.

**Comment by aidan-bowman** [comment]
> @cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.

I'll take a look. Without actually looking at the code, my best guess is either
- the cage is layered under 173 and you're just putting both sprites in the same area, OR
- the cage has a special sprite for when 173 is in the cage, and 173 just doesn't show at all (like a locker)

**Comment by TheDarkElites** [comment]
>the cage is layered under 173 and you're just putting both sprites in the same area, OR
the cage has a special sprite for when 173 is in the cage, and 173 just doesn't show at all (like a locker)

Unfortunately, after a long time of testing, it's neither. Layers are appropriately separated, and there is no special sprite. My best guess is the overlay/underlay system doesn't take layering into account.

**Comment by Tsurupeta** [comment]
> @cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.

Planes. Mob plane is higher than object plane.

**Comment by TheDarkElites** [comment]
> Planes. Mob plane is higher than object plane.

Would changing the icon layer not fix this? Or are layers relative to planes?

**Comment by Tsurupeta** [comment]
> > Planes. Mob plane is higher than object plane.
> 
> Would changing the icon layer not fix this? Or are layers relative to planes?

Second.

**Comment by TheDarkElites** [comment]
In that case, im just gonna re-sprite the cage to contain 173. It would look better that way anyhow.

**Comment by aidan-bowman** [comment]
Found the issue. Moving 173 into the cage didn't actually call the relevant code, since it all operates on the stored value, which is left unset.

**Comment by aidan-bowman** [comment]
> > @cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.
> 
> Planes. Mob plane is higher than object plane.

Underlayer is lower than mob and object, plus cage is set to mob_layer + 0.05

**Comment by aidan-bowman** [comment]
> In that case, im just gonna re-sprite the cage to contain 173. It would look better that way anyhow.

Probably not the greatest solution, makes it harder to maintain 173's sprite and also doesn't allow for him to have multiple icon states

**Comment by TheDarkElites** [comment]
>Probably not the greatest solution, makes it harder to maintain 173's sprite and also doesn't allow for him to have multiple icon states

Greatest? Probably not, however literally the only solution as of right now. I've spent all morning changing the layers around as you are suggesting right now to get it to work. Fundamentally it just doesn't.

Also, as far as the icon states, if you're referring to facing various directions, then don't worry im spriting that in too.

This is absolutely not the best idea, but there's no real alternative (unless yall have ideas), and since 173 already got his resprite I severely doubt it's gonna get changed.

Ill leave in a TODO for a proper implementation later and leave in the clean empty cages in the cage.dmi to make it easier for whoever reworks it.

**Comment by aidan-bowman** [comment]
> > Probably not the greatest solution, makes it harder to maintain 173's sprite and also doesn't allow for him to have multiple icon states
> 
> Greatest? Probably not, however literally the only solution as of right now. I've spent all morning changing the layers around as you are suggesting right now to get it to work. Fundamentally it just doesn't.
> 
> Also, as far as the icon states, if you're referring to facing various directions, then don't worry im spriting that in too.
> 
> This is absolutely not the best idea, but there's no real alternative (unless yall have ideas), and since 173 already got his resprite I severely doubt it's gonna get changed.
> 
> Ill leave in a TODO for a proper implementation later and leave in the clean empty cages in the cage.dmi to make it easier for whoever reworks it.

Lmao it's not working because you don't actually apply to the function to 173, you need to do S = dropping in the function that moves 173 into the cage because right now the function isn't actually working. Check the review changes I posted.

**Comment by TheDarkElites** [comment]
>Lmao it's not working because you don't actually apply to the function to 173, you need to do S = dropping in the function that moves 173 into the cage because right now the function isn't actually working. Check the review changes I posted.

I don't push stuff till I'm done with it..... The code you see for anything in cage rn isn't done.

For example, I already have S set to the dropping it just isn't reflected in the most recent push because _I haven't pushed it yet_

**Comment by aidan-bowman** [comment]
> > Lmao it's not working because you don't actually apply to the function to 173, you need to do S = dropping in the function that moves 173 into the cage because right now the function isn't actually working. Check the review changes I posted.
> 
> I don't push stuff till I'm done with it..... The code you see for anything in cage rn isn't done.
> 
> For example, I already have S set to the dropping it just isn't reflected in the most recent push because _I haven't pushed it yet_

Gotya

**Comment by aidan-bowman** [comment]
Perhaps setting the icon resets the underlayer?

**Comment by TheDarkElites** [comment]
> Perhaps setting the icon resets the underlayer?

Even if it did it wouldn't matter as I set the underlayer after setting the icon, plus if that was the case 173 wouldn't show up in the cell at all during testing

**Comment by aidan-bowman** [comment]
With the way the code works right now, you could probably have the cage inherit from lockers instead of structures.

**Comment by TheDarkElites** [comment]
You can change the plane of something! I never thought of that!

**Comment by aidan-bowman** [comment]
Not much of a reason to keep depreciated sprites, it's not like we can't just access older versions of files.

File: code/modules/SCP/SCPs/SCP-173.dm [hunk 5]
```diff
@@ -343,55 +361,131 @@ GLOBAL_LIST_EMPTY(scp173s)
...
+	// Max damage state before breaking out
+	var/damage_max = 5
+	// Damage state of cage
+	var/damage_state = 0
+	// Currently held 173
+	var/mob/living/scp_173/S
+
+
+/obj/structure/scp173_cage/MouseDrop_T(atom/movable/dropping, mob/user) // When someone drags 173 onto the cage
+	if(damage_state < damage_max)
```
Comment (Tsurupeta) [inline code/modules/SCP/SCPs/SCP-173.dm]: early return here

File: code/modules/SCP/SCPs/SCP-173.dm [hunk 6]
```diff
@@ -343,55 +361,131 @@ GLOBAL_LIST_EMPTY(scp173s)
...
-			name = "SCP-173 Cage"
-			playsound(loc, 'sound/machines/bolts_down.ogg', 50, 1)
-			return TRUE
+	plane = OBJ_PLANE
+	// Max damage state before breaking out
+	var/damage_max = 5
+	// Damage state of cage
+	var/damage_state = 0
+	// Currently held 173
+	var/mob/living/scp_173/S
```
Comment (Tsurupeta) [inline code/modules/SCP/SCPs/SCP-173.dm]: Don't use single fucking character, especially uppercase, for stored SCP.

**Comment by Tsurupeta** [comment]
I don't see you nulling variables of cage and 173 in their Destroy

**Comment by TheDarkElites** [comment]
fixed conflicts, now working on actual review fixes

**Review by Tsurupeta — APPROVED** [review]

**Pull request merged**